torch/distributed/_shard/checkpoint/default_planner.py

```diff
@@ -0,0 +1,346 @@
+                )
+            )
+    return read_items
+
+def create_default_metadata_only_plan(state_dict: STATE_DICT_TYPE) -> SavePlan:
+    requests = []
+    for fqn, obj in state_dict.items():
+        if isinstance(obj, ShardedTensor):
+            for shard_md in obj.metadata().shards_metadata:
+                requests.append(_create_write_item_for_shard(fqn, obj, shard_md))
+        elif isinstance(obj, torch.Tensor):
+            requests.append(_create_write_item_for_tensor(fqn, obj))
+        else:
+            requests.append(_create_write_item_for_bytesio(fqn, obj))
+    return SavePlan(requests)
+
+def create_default_local_metadata(state_dict: STATE_DICT_TYPE) -> Metadata:
+    plan = create_default_metadata_only_plan(state_dict)
+    _, md = create_default_global_save_plan([plan])
+    return md
+
+def create_write_items(fqn: str, object: Any) -> List[WriteItem]:
+    if isinstance(object, ShardedTensor):
+        return [_create_write_item_for_shard(fqn, object, shard.metadata) for shard in object.local_shards()]
+    elif isinstance(object, torch.Tensor):
+        return [_create_write_item_for_tensor(fqn, object)]
+    else:
+        return [_create_write_item_for_bytesio(fqn, object)]
+
+def create_read_items(fqn: str, md: STORAGE_TYPES, obj: Any) -> List[ReadItem]:
+    if isinstance(md, BytesStorageMetadata):
+        return [_create_read_item_for_byteio(
+            dest_index=MetadataIndex(fqn),
+            dest_offset=0,
+            storage_index=MetadataIndex(fqn),
+            storage_offset=0,
+            length=0
+        )]
+    elif isinstance(obj, ShardedTensor):
+        local_shards = obj.local_shards()
+    elif isinstance(obj, torch.Tensor):
+        local_shards = [_create_shard_from_tensor(obj)]
+    else:
+        raise ValueError(
+            f"Invalid checkpoint metadata for {fqn}, " +
+            f"expected BytesStorageMetadata but found {type(md)}"
+        )
+
+    return _create_sharded_read_items(
+        fqn,
+        md,
+        local_shards)
+
+def create_default_local_load_plan(
+    state_dict: Dict[str, Any],
+    metadata: Metadata,
+) -> LoadPlan:
+    requests = []
+
+    """
+    Use the loaded metadata and the current state dict to map the saved tensors to current tensor
+    """
+    for fqn, obj in state_dict.items():
+        md = metadata.state_dict_metadata[fqn]
+        requests += create_read_items(fqn, md, obj)
+
+    return LoadPlan(requests)
+
+def create_default_global_load_plan(all_plans: List[LoadPlan]) -> List[LoadPlan]:
+    return all_plans
+
+def create_default_local_save_plan(state_dict: Dict[str, Any], is_coordinator: bool):
+    requests = []
+    for fqn, obj in state_dict.items():
+        if isinstance(obj, ShardedTensor) or is_coordinator:
+            requests += create_write_items(fqn, obj)
+    return SavePlan(requests)
+
+def create_default_global_save_plan(all_plans: List[SavePlan]) -> Tuple[List[SavePlan], Metadata]:
```

Comment:
My general feeling about this file: there're way too much apis and helper functions mixed up together, and it's very hard to track each api and often forgot what the previous api is. question: are those all apis we are going to expose to the user? can we move out some of the apis/helpers out if it's not exposing to the user, especially some helper functions might be useful for other custom created planners?

Reply:
That's true, there's a huge mix between the default planners and the helpers.

The issue of how much we want to expose is kinda tricky.

As I've explored using the planner API, the more I've ran into scenarios were I implement them in terms of those helper primitives.

I'll move the helpers out to `planner_helpers.py` and only keep the following 5 as public as they've shown to be quite useful so far:

```
create_default_local_save_plan
create_default_global_load_plan
create_default_global_save_plan
```

The following two I'll keep around, but make them private, as their value is mostly testing:

```
create_default_local_metadata
create_default_metadata_only_plan
```